File: resources/logdb/createdb.py
import sqlite3
import contextlib
from os import path
import sys

SHOW_PROGRESS = False

def classAndPackage(fqcn):
    lastdot = fqcn.rfind('.')
    classname = fqcn[lastdot + 1:]
    package = fqcn[:lastdot]
    return classname, package

def quotOrNull(txt):
    if txt == None:
        return "NULL"
    else:
        return "'" + txt + "'"

chunksize = 10000;

def add_chunk_of_log_enties(chunk, cursor):
    cursor.execute("begin transaction")
    result = cursor.executemany("insert into log values(?,?)", chunk)
    cursor.execute("commit")
    return len(chunk)

def initialize_log_table(infile, connection):
    print("Initializing log table from {} ...".format(infile))
    total = 0
    with contextlib.closing(connection.cursor()) as cursor:
        # truncate log table
        cursor.execute("delete from log")
        with open(infile, 'r') as inlog:
            chunk = []
            for idx, entry in enumerate(inlog):
                chunk.append((idx + 1, entry.rstrip("\n")))
                if idx % chunksize == chunksize - 1:
                    total += add_chunk_of_log_enties(chunk, cursor)
                    if SHOW_PROGRESS:
                        print(".", end ="")
                    chunk.clear()
            total += add_chunk_of_log_enties(chunk, cursor)
    if SHOW_PROGRESS:
        print("\n")
    print("Added {} rows to table {}".format(str(total), "log"))

def create_log_table(connection):
    with contextlib.closing(connection.cursor()) as cursor:
        cursor.execute("create table if not exists log(line integer primary key, entry text)")

def connectdb(dbname):
    return sqlite3.connect(dbname, isolation_level=None)

def create_log_db(log_path, dbname):
    connection = connectdb(dbname)
    create_log_table(connection)
    initialize_log_table(log_path, connection)
    return connection

log_dir = "data"

def main(logname):
    debug_log = path.join(log_dir, logname + ".err")
    debug_db = path.join("", logname + ".db")
    return create_log_db(debug_log, debug_db)

if __name__ == "__main__":
    if len(sys.argv) > 1:
        logname = sys.argv[1]
    else:
        raise Exception("Missing command line parameter for log name.")
    main(logname)

File: loghog/src/main/java/com/contrastsecurity/agent/loghog/db/CreateDb.java
package com.contrastsecurity.agent.loghog.db;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class CreateDb {
    private static final boolean SHOW_PROGRESS = false;
    private static final int CHUNK_SIZE = 10000;

    public static String[] classAndPackage(String fqcn) {
        int lastDot = fqcn.lastIndexOf('.');
        String classname = fqcn.substring(lastDot + 1);
        String packageName = fqcn.substring(0, lastDot);
        return new String[]{classname, packageName};
    }

    public static String quotOrNull(String txt) {
        if (txt == null) {
            return "NULL";
        } else {
            return "'" + txt + "'";
        }
    }

    public static int addChunkOfLogEntries(List<String[]> chunk, Connection connection) throws SQLException {
        String sql = "INSERT INTO log VALUES (?, ?)";
        try (PreparedStatement pstmt = connection.prepareStatement(sql)) {
            connection.setAutoCommit(false);
            for (String[] entry : chunk) {
                pstmt.setInt(1, Integer.parseInt(entry[0]));
                pstmt.setString(2, entry[1]);
                pstmt.addBatch();
            }
            int[] result = pstmt.executeBatch();
            connection.commit();
            return result.length;
        }
    }

    public static void initializeLogTable(String infile, Connection connection) throws IOException, SQLException {
        System.out.println("Initializing log table from " + infile + " ...");
        int total = 0;
        try (Statement stmt = connection.createStatement()) {
            stmt.execute("DELETE FROM log");
        }
        try (BufferedReader inlog = new BufferedReader(new FileReader(infile))) {
            List<String[]> chunk = new ArrayList<>();
            String entry;
            int idx = 0;
            while ((entry = inlog.readLine()) != null) {
                chunk.add(new String[]{String.valueOf(idx + 1), entry});
                if (idx % CHUNK_SIZE == CHUNK_SIZE - 1) {
                    total += addChunkOfLogEntries(chunk, connection);
                    if (SHOW_PROGRESS) {
                        System.out.print(".");
                    }
                    chunk.clear();
                }
                idx++;
            }
            total += addChunkOfLogEntries(chunk, connection);
        }
        if (SHOW_PROGRESS) {
            System.out.println();
        }
        System.out.println("Added " + total + " rows to table log");
    }

    public static void createLogTable(Connection connection) throws SQLException {
        String sql = "CREATE TABLE IF NOT EXISTS log (line INTEGER PRIMARY KEY, entry TEXT)";
        try (Statement stmt = connection.createStatement()) {
            stmt.execute(sql);
        }
    }

    public static Connection connectDb(String dbname) throws SQLException {
        return DriverManager.getConnection("jdbc:sqlite:" + dbname);
    }

    public static Connection createLogDb(String logPath, String dbname) throws IOException, SQLException {
        Connection connection = connectDb(dbname);
        createLogTable(connection);
        initializeLogTable(logPath, connection);
        return connection;
    }

    public static void main(String[] args) {
        String logname = "sb-san-petclinic";
        if (args.length > 0) {
            logname = args[0];
        }
        String debugLog = logname + ".err";
        String debugDb = logname + ".db";
        try {
            createLogDb(debugLog, debugDb);
        } catch (IOException | SQLException e) {
            e.printStackTrace();
        }
    }
}